mesonbuild/build.py

```diff
@@ -320,16 +303,52 @@ def process_sourcelist(self, sources):
                 msg = 'Bad source of type {!r} in target {!r}.'.format(type(s).__name__, self.name)
                 raise InvalidArguments(msg)
 
+    @staticmethod
+    def can_compile_remove_sources(compiler, sources):
+        removed = False
+        for s in sources:
+            if compiler.can_compile(s):
+                sources.remove(s)
```

Comment:
I already changed that to iterate over a copy and edit the original list. Will push it soon.